service/history/replication/raw_task_converter.go

```diff
@@ -632,112 +617,167 @@ func convertGetHistoryError(
-		replicationCache: replicationCache,
-		targetClusterID:  targetClusterID,
-		executionManager: executionManager,
-		mutableState:     mutableState,
-		logger: log.NewLazyLogger(
-			logger,
-			func() []tag.Tag {
-				return []tag.Tag{
-					tag.WorkflowNamespaceID(taskInfo.NamespaceID),
-					tag.WorkflowID(taskInfo.WorkflowID),
-					tag.WorkflowRunID(taskInfo.RunID),
-					tag.TaskKey(taskInfo.GetKey()),
-					tag.Task(taskInfo),
-				}
-			},
-		),
+		shardContext:       shardContext,
+		workflowCache:      workflowCache,
+		eventBlobCache:     eventBlobCache,
+		replicationCache:   replicationCache,
+		executionManager:   executionManager,
+		syncStateRetriever: syncStateRetriever,
+		shardID:            shardContext.GetShardID(),
+		logger:             logger,
 	}
 }
 
-func (c *syncVersionedTransitionTaskConverter) convert() (*replicationspb.ReplicationTask, error) {
+func (c *syncVersionedTransitionTaskConverter) convert(
+	ctx context.Context,
+	taskInfo *tasks.SyncVersionedTransitionTask,
+	targetClusterID int32,
+	mutableState workflow.MutableState,
+) (*replicationspb.ReplicationTask, error) {
+	executionInfo := mutableState.GetExecutionInfo()
 
 	// If workflow is not on any versionedTransition (in an unknown state from state-based replication perspective),
 	// we can't convert this raw task to a replication task, instead we need to rely on its task equivalents.
-	if len(c.mutableState.GetExecutionInfo().TransitionHistory) == 0 {
-		return c.convertTaskEquivalents()
+	if len(executionInfo.TransitionHistory) == 0 {
+		return c.convertTaskEquivalents(ctx, taskInfo, targetClusterID)
 	}
 
-	progress := c.replicationCache.Get(c.taskInfo.RunID, c.targetClusterID)
-	if progress.VersionedTransitionSent(c.taskInfo.VersionedTransition) {
-		return c.generateVerifyVersionedTransitionTask()
+	progress := c.replicationCache.Get(taskInfo.RunID, targetClusterID)
+
+	if progress.VersionedTransitionSent(taskInfo.VersionedTransition) {
+		return c.generateVerifyVersionedTransitionTask(taskInfo, mutableState)
 	}
 
-	if !c.onCurrentBranch(c.taskInfo.VersionedTransition) {
-		if c.taskInfo.FirstEventID == common.EmptyEventID && c.taskInfo.NextEventID == common.EmptyEventID && len(c.taskInfo.NewRunID) == 0 {
+	if !c.onCurrentBranch(mutableState, taskInfo.VersionedTransition) {
+		if taskInfo.FirstEventID == common.EmptyEventID && taskInfo.NextEventID == common.EmptyEventID && len(taskInfo.NewRunID) == 0 {
 			return nil, nil
 		}
-		return c.generateBackfillHistoryTask()
+		return c.generateBackfillHistoryTask(ctx, taskInfo, targetClusterID)
+	}
+
+	var targetHistoryItems [][]*historyspb.VersionHistoryItem
+	if progress != nil {
+		targetHistoryItems = progress.eventVersionHistoryItems
 	}
 
-	// lastSyncedTransition := progress.LastSyncedTransition()
-	// TODO: SyncWorkflowStateMutation & SyncWorkflowStateSnapshot based on lastSyncedTransition
-	return nil, nil
+	result, err := c.syncStateRetriever.GetSyncWorkflowStateArtifactFromMutableState(
```

Comment:
Next refactor will be: release mutable state before reading history events, so the active side traffic will not be blocked.